ocaml/xapi/server_helpers.ml

```diff
@@ -118,26 +128,44 @@ let dispatch_exn_wrapper f =
   with exn -> let code, params = ExnHelper.error_of_exn exn in API.response_of_failure code params
 
 let do_dispatch ?session_id ?forward_op ?self called_async supports_async called_fn_name op_fn
-    marshaller_fn fd http_req label generate_task_for =
+    marshaller_fn fd http_req label sync_ty generate_task_for =
 
   if (called_async && (not supports_async))
   then API.response_of_fault ("No async mode for this operation (rpc: "^called_fn_name^")")
   else
     let __context = Context.of_http_req ?session_id ~generate_task_for ~supports_async ~label ~http_req ~fd in
-    if called_async
-    then begin
+    let sync () =
+      exec_with_context ~__context ~called_async ?f_forward:forward_op ~marshaller:marshaller_fn op_fn |>
+      marshaller_fn |>
+      Rpc.success
+    in
+    let default_async () =
       (* Fork thread in which to execute async call *)
       ignore (Thread.create
                 (fun () ->
                    exec_with_context ~__context ~called_async ?f_forward:forward_op ~marshaller:marshaller_fn op_fn)
                 ());
       (* Return task id immediately *)
       Rpc.success (API.rpc_of_ref_task (Context.get_task_id __context))
-    end else
-      let result =
-        exec_with_context ~__context ~called_async ?f_forward:forward_op ~marshaller:marshaller_fn op_fn
-      in
-      Rpc.success (marshaller_fn result)
+    in
+    let internal_async () =
+      (** similar to `default_async`, except we are zealous when completing the task
+        * this should be used when converting sync xapi <-> xapi calls to async *)
+      ignore (Thread.create
+                (fun () ->
+                  try
+                    let res = exec_with_context ~__context ~called_async ?f_forward:forward_op ~marshaller:marshaller_fn op_fn |> marshaller_fn in
+                    TaskHelper.complete ~__context (Some res)
```

Comment:
I think that we should leave `exec_with_context` be the only place were tasks are completed. This function already has a notion of `need_complete` inside of it, which we could pull out.